CcrRepositoryIT.testIndividualActionsTimeout failing

Failing on intake-master
`
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1648/console
`

but unfortunately does not reproduce for me 
```
./gradlew :x-pack:plugin:ccr:internalClusterTest \
  -Dtests.seed=75F3403255B8F3AF \
  -Dtests.class=org.elasticsearch.xpack.ccr.CcrRepositoryIT \
  -Dtests.method="testIndividualActionsTimeout" \
  -Dtests.security.manager=true \
  -Dtests.locale=ar-AE \
  -Dtests.timezone=Europe/Saratov \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

relates https://github.com/elastic/elasticsearch/issues/37887

assertion failure log
```
java.lang.AssertionError: expected:<0> but was:<1>
	at __randomizedtesting.SeedInfo.seed([75F3403255B8F3AF:11EEACEEE6CAC19C]:0)
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.elasticsearch.xpack.ccr.CcrRepositoryIT.testIndividualActionsTimeout(CcrRepositoryIT.java:370)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
```

Seems like this fails less often but there are still failures 

## Example build failure
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=ubuntu-14.04&&virtual/215/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1684/console

## Reproduction line 
does not reproduce locally
```
./gradlew :x-pack:plugin:ccr:internalClusterTest -Dtests.seed=36B1DA4D705F0B8F -Dtests.class=org.elasticsearch.xpack.ccr.CcrRepositoryIT -Dtests.method="testIndividualActionsTimeout" -Dtests.security.manager=true -Dtests.locale=mt -Dtests.timezone=Atlantic/Madeira -Dcompiler.java=11 -Druntime.java=8
```

## Example relevant log:
```
11:26:45   1> org.apache.lucene.util.SetOnce$AlreadySetException: The object cannot be set twice!
11:26:45   1> 	at org.apache.lucene.util.SetOnce.set(SetOnce.java:69) ~[lucene-core-8.0.0-snapshot-83f9835.jar:8.0.0-snapshot-83f9835 83f9835a47a00a2ec58a4cf5fc0d492497cf7898 - jpountz - 2019-01-21 13:06:00]
11:26:45   1> 	at org.elasticsearch.common.logging.NodeAndClusterIdConverter.setNodeIdAndClusterId(NodeAndClusterIdConverter.java:59) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.common.logging.NodeAndClusterIdStateListener.onNewClusterState(NodeAndClusterIdStateListener.java:69) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onNewClusterState(ClusterStateObserver.java:308) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.clusterChanged(ClusterStateObserver.java:193) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.cluster.service.ClusterApplierService.lambda$callClusterStateListeners$6(ClusterApplierService.java:481) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at java.util.concurrent.ConcurrentHashMap$KeySpliterator.forEachRemaining(ConcurrentHashMap.java:3527) [?:1.8.0_202]
11:26:45   1> 	at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:743) [?:1.8.0_202]
11:26:45   1> 	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580) [?:1.8.0_202]
11:26:45   1> 	at org.elasticsearch.cluster.service.ClusterApplierService.callClusterStateListeners(ClusterApplierService.java:478) [elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:467) [elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:414) [elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:165) [elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681) [elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
11:26:45   1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_202]
11:26:45   1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_202]
11:26:45   1> 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_202]


11:26:45 FAILURE 14.9s J3 | CcrRepositoryIT.testIndividualActionsTimeout <<< FAILURES!
11:26:45    > Throwable #1: java.lang.AssertionError: 
11:26:45    > Expected: a value greater than <0>
11:26:45    >      but: <0> was equal to <0>
11:26:45    > 	at __randomizedtesting.SeedInfo.seed([36B1DA4D705F0B8F:52AC3691C32D39BC]:0)
11:26:45    > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
11:26:45    > 	at org.elasticsearch.xpack.ccr.CcrRepositoryIT.testIndividualActionsTimeout(CcrRepositoryIT.java:372)
11:26:45    > 	at java.lang.Thread.run(Thread.java:748)
```

## Frequency

6 time Today

--- a/server/src/main/java/org/elasticsearch/action/support/ListenerTimeouts.java
+++ b/server/src/main/java/org/elasticsearch/action/support/ListenerTimeouts.java
@@ -1,0 +1,77 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.ElasticsearchTimeoutException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class ListenerTimeouts {
+
+    public static <Response> ActionListener<Response> wrapWithTimeout(ThreadPool threadPool, ActionListener<Response> listener,
+                                                                      TimeValue timeout, String listenerName) {
+        TimeoutableListener<Response> wrappedListener = new TimeoutableListener<>(listener, timeout, listenerName);
+        wrappedListener.cancellable = threadPool.schedule(wrappedListener, timeout, ThreadPool.Names.GENERIC);
+        return wrappedListener;
+    }
+
+    private static class TimeoutableListener<Response> implements ActionListener<Response>, Runnable {
+
+        private final AtomicBoolean isDone = new AtomicBoolean(false);
+        private final ActionListener<Response> delegate;
+        private final TimeValue timeout;
+        private final String listenerName;
+        private volatile Scheduler.ScheduledCancellable cancellable;
+
+        private TimeoutableListener(ActionListener<Response> delegate, TimeValue timeout, String listenerName) {
+            this.delegate = delegate;
+            this.timeout = timeout;
+            this.listenerName = listenerName;
+        }
+
+        @Override
+        public void onResponse(Response response) {
+            if (isDone.compareAndSet(false, true)) {
+                cancellable.cancel();
+                delegate.onResponse(response);
+            }
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            if (isDone.compareAndSet(false, true)) {
+                cancellable.cancel();
+                delegate.onFailure(e);
+            }
+        }
+
+        @Override
+        public void run() {
+            if (isDone.compareAndSet(false, true)) {
+                String timeoutMessage = "[" + listenerName + "]" + " timed out after [" + timeout + "]";
+                delegate.onFailure(new ElasticsearchTimeoutException(timeoutMessage));
+            }
+        }
+    }
+}

--- a/server/src/main/java/org/elasticsearch/action/support/ListenerTimeouts.java
+++ b/server/src/main/java/org/elasticsearch/action/support/ListenerTimeouts.java
@@ -29,6 +29,17 @@ import java.util.concurrent.atomic.AtomicBoolean;
 
 public class ListenerTimeouts {
 
+    /**
+     * Wraps a listener with a listener that can timeout. After the timeout period the
+     * {@link ActionListener#onFailure(Exception)} will be called with a
+     * {@link ElasticsearchTimeoutException} if the listener has not already been completed.
+     *
+     * @param threadPool used to schedule the timeout
+     * @param listener to that can timeout
+     * @param timeout period before listener failed
+     * @param listenerName name of the listener for timeout exception
+     * @return the wrapped listener that will timeout
+     */
     public static <Response> ActionListener<Response> wrapWithTimeout(ThreadPool threadPool, ActionListener<Response> listener,
                                                                       TimeValue timeout, String listenerName) {
         TimeoutableListener<Response> wrappedListener = new TimeoutableListener<>(listener, timeout, listenerName);

--- a/server/src/main/java/org/elasticsearch/action/support/ListenerTimeouts.java
+++ b/server/src/main/java/org/elasticsearch/action/support/ListenerTimeouts.java
@@ -37,13 +37,14 @@ public class ListenerTimeouts {
      * @param threadPool used to schedule the timeout
      * @param listener to that can timeout
      * @param timeout period before listener failed
+     * @param executor to use for scheduling timeout
      * @param listenerName name of the listener for timeout exception
      * @return the wrapped listener that will timeout
      */
-    public static <Response> ActionListener<Response> wrapWithTimeout(ThreadPool threadPool, ActionListener<Response> listener,
-                                                                      TimeValue timeout, String listenerName) {
+    public static <Response> ActionListener<Response> wrapWithTimeout(ThreadPool threadPool,  ActionListener<Response> listener,
+                                                                      TimeValue timeout, String executor, String listenerName) {
         TimeoutableListener<Response> wrappedListener = new TimeoutableListener<>(listener, timeout, listenerName);
-        wrappedListener.cancellable = threadPool.schedule(wrappedListener, timeout, ThreadPool.Names.GENERIC);
+        wrappedListener.cancellable = threadPool.schedule(wrappedListener, timeout, executor);
         return wrappedListener;
     }
 

--- a/server/src/test/java/org/elasticsearch/action/support/ListenerTimeoutsTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/ListenerTimeoutsTests.java
@@ -49,9 +49,9 @@ public class ListenerTimeoutsTests extends ESTestCase {
     }
 
     public void testListenerTimeout() {
-        AtomicBoolean completed = new AtomicBoolean(false);
+        AtomicBoolean success = new AtomicBoolean(false);
         AtomicReference<Exception> exception = new AtomicReference<>();
-        ActionListener<Void> listener = wrap(completed, exception);
+        ActionListener<Void> listener = wrap(success, exception);
 
         ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, generic, "test");
         assertTrue(taskQueue.hasDeferredTasks());
@@ -61,14 +61,14 @@ public class ListenerTimeoutsTests extends ESTestCase {
         wrapped.onResponse(null);
         wrapped.onFailure(new IOException("incorrect exception"));
 
-        assertFalse(completed.get());
+        assertFalse(success.get());
         assertThat(exception.get(), instanceOf(ElasticsearchTimeoutException.class));
     }
 
     public void testFinishNormallyBeforeTimeout() {
-        AtomicBoolean completed = new AtomicBoolean(false);
+        AtomicBoolean success = new AtomicBoolean(false);
         AtomicReference<Exception> exception = new AtomicReference<>();
-        ActionListener<Void> listener = wrap(completed, exception);
+        ActionListener<Void> listener = wrap(success, exception);
 
         ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, generic, "test");
         wrapped.onResponse(null);
@@ -79,14 +79,14 @@ public class ListenerTimeoutsTests extends ESTestCase {
         taskQueue.advanceTime();
         taskQueue.runAllRunnableTasks();
 
-        assertTrue(completed.get());
+        assertTrue(success.get());
         assertNull(exception.get());
     }
 
     public void testFinishExceptionallyBeforeTimeout() {
-        AtomicBoolean completed = new AtomicBoolean(false);
+        AtomicBoolean success = new AtomicBoolean(false);
         AtomicReference<Exception> exception = new AtomicReference<>();
-        ActionListener<Void> listener = wrap(completed, exception);
+        ActionListener<Void> listener = wrap(success, exception);
 
         ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, generic, "test");
         wrapped.onFailure(new IOException("boom"));
@@ -95,22 +95,25 @@ public class ListenerTimeoutsTests extends ESTestCase {
         taskQueue.advanceTime();
         taskQueue.runAllRunnableTasks();
 
-        assertFalse(completed.get());
+        assertFalse(success.get());
         assertThat(exception.get(), instanceOf(IOException.class));
     }
 
-    private ActionListener<Void> wrap(AtomicBoolean completed, AtomicReference<Exception> exception) {
+    private ActionListener<Void> wrap(AtomicBoolean success, AtomicReference<Exception> exception) {
         return new ActionListener<Void>() {
+
+            private final AtomicBoolean completed = new AtomicBoolean();
+
             @Override
             public void onResponse(Void aVoid) {
-                assert completed.get() == false : "Should not be called twice";
-                completed.set(true);
+                assertTrue(completed.compareAndSet(false, true));
+                assertTrue(success.compareAndSet(false, true));
             }
 
             @Override
             public void onFailure(Exception e) {
-                assert exception.get() == null : "Should not be called twice";
-                exception.set(e);
+                assertTrue(completed.compareAndSet(false, true));
+                assertTrue(exception.compareAndSet(null, e));
             }
         };
     }

--- a/server/src/test/java/org/elasticsearch/action/support/ListenerTimeoutsTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/ListenerTimeoutsTests.java
@@ -1,0 +1,93 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.ElasticsearchTimeoutException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.coordination.DeterministicTaskQueue;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Before;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.elasticsearch.node.Node.NODE_NAME_SETTING;
+import static org.hamcrest.core.IsInstanceOf.instanceOf;
+
+public class ListenerTimeoutsTests extends ESTestCase {
+
+    private final TimeValue timeout = TimeValue.timeValueMillis(10);
+    private DeterministicTaskQueue taskQueue;
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        Settings settings = Settings.builder().put(NODE_NAME_SETTING.getKey(), "node").build();
+        taskQueue = new DeterministicTaskQueue(settings, random());
+    }
+
+    public void testListenerTimeout() {
+        AtomicBoolean completed = new AtomicBoolean(false);
+        AtomicReference<Exception> exception = new AtomicReference<>();
+        ActionListener<Void> listener = wrap(completed, exception);
+
+        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, "test");
+        assertTrue(taskQueue.hasDeferredTasks());
+        taskQueue.advanceTime();
+        taskQueue.runAllRunnableTasks();
+
+        wrapped.onResponse(null);
+
+        assertFalse(completed.get());
+        assertThat(exception.get(), instanceOf(ElasticsearchTimeoutException.class));
+    }
+
+    public void testFinishBeforeTimeout() {
+        AtomicBoolean completed = new AtomicBoolean(false);
+        AtomicReference<Exception> exception = new AtomicReference<>();
+        ActionListener<Void> listener = wrap(completed, exception);
+
+        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, "test");
+        wrapped.onResponse(null);
+
+        assertTrue(taskQueue.hasDeferredTasks());
+        taskQueue.advanceTime();
+        taskQueue.runAllRunnableTasks();
+
+        assertTrue(completed.get());
+        assertNull(exception.get());
+    }
+
+    private ActionListener<Void> wrap(AtomicBoolean completed, AtomicReference<Exception> exception) {
+        return new ActionListener<Void>() {
+            @Override
+            public void onResponse(Void aVoid) {
+                completed.set(true);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                exception.set(e);
+            }
+        };
+    }
+}

--- a/server/src/test/java/org/elasticsearch/action/support/ListenerTimeoutsTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/ListenerTimeoutsTests.java
@@ -25,8 +25,10 @@ import org.elasticsearch.cluster.coordination.DeterministicTaskQueue;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.ThreadPool;
 import org.junit.Before;
 
+import java.io.IOException;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 
@@ -36,6 +38,7 @@ import static org.hamcrest.core.IsInstanceOf.instanceOf;
 public class ListenerTimeoutsTests extends ESTestCase {
 
     private final TimeValue timeout = TimeValue.timeValueMillis(10);
+    private final String generic = ThreadPool.Names.GENERIC;
     private DeterministicTaskQueue taskQueue;
 
     @Before
@@ -50,23 +53,26 @@ public class ListenerTimeoutsTests extends ESTestCase {
         AtomicReference<Exception> exception = new AtomicReference<>();
         ActionListener<Void> listener = wrap(completed, exception);
 
-        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, "test");
+        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, generic, "test");
         assertTrue(taskQueue.hasDeferredTasks());
         taskQueue.advanceTime();
         taskQueue.runAllRunnableTasks();
 
         wrapped.onResponse(null);
+        wrapped.onFailure(new IOException("incorrect exception"));
 
         assertFalse(completed.get());
         assertThat(exception.get(), instanceOf(ElasticsearchTimeoutException.class));
     }
 
-    public void testFinishBeforeTimeout() {
+    public void testFinishNormallyBeforeTimeout() {
         AtomicBoolean completed = new AtomicBoolean(false);
         AtomicReference<Exception> exception = new AtomicReference<>();
         ActionListener<Void> listener = wrap(completed, exception);
 
-        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, "test");
+        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, generic, "test");
+        wrapped.onResponse(null);
+        wrapped.onFailure(new IOException("boom"));
         wrapped.onResponse(null);
 
         assertTrue(taskQueue.hasDeferredTasks());
@@ -77,15 +83,33 @@ public class ListenerTimeoutsTests extends ESTestCase {
         assertNull(exception.get());
     }
 
+    public void testFinishExceptionallyBeforeTimeout() {
+        AtomicBoolean completed = new AtomicBoolean(false);
+        AtomicReference<Exception> exception = new AtomicReference<>();
+        ActionListener<Void> listener = wrap(completed, exception);
+
+        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, generic, "test");
+        wrapped.onFailure(new IOException("boom"));
+
+        assertTrue(taskQueue.hasDeferredTasks());
+        taskQueue.advanceTime();
+        taskQueue.runAllRunnableTasks();
+
+        assertFalse(completed.get());
+        assertThat(exception.get(), instanceOf(IOException.class));
+    }
+
     private ActionListener<Void> wrap(AtomicBoolean completed, AtomicReference<Exception> exception) {
         return new ActionListener<Void>() {
             @Override
             public void onResponse(Void aVoid) {
+                assert completed.get() == false : "Should not be called twice";
                 completed.set(true);
             }
 
             @Override
             public void onFailure(Exception e) {
+                assert exception.get() == null : "Should not be called twice";
                 exception.set(e);
             }
         };

--- a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java
+++ b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java
@@ -494,7 +494,7 @@ public class CcrRepository extends AbstractLifecycleComponent implements Reposit
                                         error.compareAndSet(null, Tuple.tuple(fileToRecover.metadata(), e));
                                         requestSeqIdTracker.markSeqNoAsCompleted(requestSeqId);
                                     }
-                                    ), timeout, GetCcrRestoreFileChunkAction.NAME);
+                                    ), timeout, ThreadPool.Names.GENERIC, GetCcrRestoreFileChunkAction.NAME);
                             remoteClient.execute(GetCcrRestoreFileChunkAction.INSTANCE, request, listener);
                         } catch (Exception e) {
                             error.compareAndSet(null, Tuple.tuple(fileToRecover.metadata(), e));

--- a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java
+++ b/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java
@@ -16,6 +16,7 @@ import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.admin.cluster.state.ClusterStateRequest;
 import org.elasticsearch.action.admin.cluster.state.ClusterStateResponse;
 import org.elasticsearch.action.admin.indices.mapping.put.PutMappingRequest;
+import org.elasticsearch.action.support.ListenerTimeouts;
 import org.elasticsearch.action.support.PlainActionFuture;
 import org.elasticsearch.client.Client;
 import org.elasticsearch.cluster.metadata.IndexMetaData;
@@ -105,7 +106,7 @@ public class CcrRepository extends AbstractLifecycleComponent implements Reposit
     private final ThreadPool threadPool;
 
     private final CounterMetric throttledTime = new CounterMetric();
-    
+
     public CcrRepository(RepositoryMetaData metadata, Client client, CcrLicenseChecker ccrLicenseChecker, Settings settings,
                          CcrSettings ccrSettings, ThreadPool threadPool) {
         this.metadata = metadata;
@@ -389,7 +390,8 @@ public class CcrRepository extends AbstractLifecycleComponent implements Reposit
         protected void restoreFiles(List<FileInfo> filesToRecover, Store store) throws IOException {
             logger.trace("[{}] starting CCR restore of {} files", shardId, filesToRecover);
 
-            try (MultiFileWriter multiFileWriter = new MultiFileWriter(store, recoveryState.getIndex(), "", logger, () -> {})) {
+            try (MultiFileWriter multiFileWriter = new MultiFileWriter(store, recoveryState.getIndex(), "", logger, () -> {
+            })) {
                 final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
                 final AtomicReference<Tuple<StoreFileMetaData, Exception>> error = new AtomicReference<>();
 
@@ -444,8 +446,9 @@ public class CcrRepository extends AbstractLifecycleComponent implements Reposit
                             logger.trace("[{}] [{}] fetching chunk for file [{}], expected offset: {}, size: {}", shardId, snapshotId,
                                 fileToRecover.name(), fileSession.lastOffset, bytesRequested);
 
-                            remoteClient.execute(GetCcrRestoreFileChunkAction.INSTANCE, request,
-                                ActionListener.wrap(
+                            TimeValue timeout = ccrSettings.getRecoveryActionTimeout();
+                            ActionListener<GetCcrRestoreFileChunkAction.GetCcrRestoreFileChunkResponse> listener =
+                                ListenerTimeouts.wrapWithTimeout(threadPool, ActionListener.wrap(
                                     r -> threadPool.generic().execute(new AbstractRunnable() {
                                         @Override
                                         public void onFailure(Exception e) {
@@ -491,7 +494,8 @@ public class CcrRepository extends AbstractLifecycleComponent implements Reposit
                                         error.compareAndSet(null, Tuple.tuple(fileToRecover.metadata(), e));
                                         requestSeqIdTracker.markSeqNoAsCompleted(requestSeqId);
                                     }
-                                ));
+                                    ), timeout, GetCcrRestoreFileChunkAction.NAME);
+                            remoteClient.execute(GetCcrRestoreFileChunkAction.INSTANCE, request, listener);
                         } catch (Exception e) {
                             error.compareAndSet(null, Tuple.tuple(fileToRecover.metadata(), e));
                             requestSeqIdTracker.markSeqNoAsCompleted(requestSeqId);

--- a/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/CcrRepositoryIT.java
+++ b/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/CcrRepositoryIT.java
@@ -37,6 +37,7 @@ import org.elasticsearch.snapshots.RestoreInfo;
 import org.elasticsearch.snapshots.RestoreService;
 import org.elasticsearch.snapshots.Snapshot;
 import org.elasticsearch.test.transport.MockTransportService;
+import org.elasticsearch.transport.TransportActionProxy;
 import org.elasticsearch.transport.TransportService;
 import org.elasticsearch.xpack.CcrIntegTestCase;
 import org.elasticsearch.xpack.ccr.action.repositories.GetCcrRestoreFileChunkAction;
@@ -325,7 +326,6 @@ public class CcrRepositoryIT extends CcrIntegTestCase {
         }
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/38027")
     public void testIndividualActionsTimeout() throws Exception {
         ClusterUpdateSettingsRequest settingsRequest = new ClusterUpdateSettingsRequest();
         TimeValue timeValue = TimeValue.timeValueMillis(100);
@@ -348,7 +348,8 @@ public class CcrRepositoryIT extends CcrIntegTestCase {
             MockTransportService mockTransportService = (MockTransportService) transportService;
             transportServices.add(mockTransportService);
             mockTransportService.addSendBehavior((connection, requestId, action, request, options) -> {
-                if (action.equals(GetCcrRestoreFileChunkAction.NAME) == false) {
+                if (action.equals(GetCcrRestoreFileChunkAction.NAME) == false &&
+                    action.equals(TransportActionProxy.getProxyAction(GetCcrRestoreFileChunkAction.NAME)) == false) {
                     connection.sendRequest(requestId, action, request, options);
                 }
             });
@@ -370,33 +371,34 @@ public class CcrRepositoryIT extends CcrIntegTestCase {
             .renameReplacement(followerIndex).masterNodeTimeout(new TimeValue(1L, TimeUnit.HOURS))
             .indexSettings(settingsBuilder);
 
-        final RestoreService restoreService = getFollowerCluster().getCurrentMasterNodeInstance(RestoreService.class);
-        final ClusterService clusterService = getFollowerCluster().getCurrentMasterNodeInstance(ClusterService.class);
-        PlainActionFuture<RestoreInfo> future = PlainActionFuture.newFuture();
-        restoreService.restoreSnapshot(restoreRequest, waitForRestore(clusterService, future));
-
-        // Depending on when the timeout occurs this can fail in two ways. If it times-out when fetching
-        // metadata this will throw an exception. If it times-out when restoring a shard, the shard will
-        // be marked as failed. Either one is a success for the purpose of this test.
         try {
-            RestoreInfo restoreInfo = future.actionGet();
-            assertThat(restoreInfo.failedShards(), greaterThan(0));
-            assertThat(restoreInfo.successfulShards(), lessThan(restoreInfo.totalShards()));
-            assertEquals(numberOfPrimaryShards, restoreInfo.totalShards());
-        } catch (Exception e) {
-            assertThat(ExceptionsHelper.unwrapCause(e), instanceOf(ElasticsearchTimeoutException.class));
+            final RestoreService restoreService = getFollowerCluster().getCurrentMasterNodeInstance(RestoreService.class);
+            final ClusterService clusterService = getFollowerCluster().getCurrentMasterNodeInstance(ClusterService.class);
+            PlainActionFuture<RestoreInfo> future = PlainActionFuture.newFuture();
+            restoreService.restoreSnapshot(restoreRequest, waitForRestore(clusterService, future));
+
+            // Depending on when the timeout occurs this can fail in two ways. If it times-out when fetching
+            // metadata this will throw an exception. If it times-out when restoring a shard, the shard will
+            // be marked as failed. Either one is a success for the purpose of this test.
+            try {
+                RestoreInfo restoreInfo = future.actionGet();
+                assertThat(restoreInfo.failedShards(), greaterThan(0));
+                assertThat(restoreInfo.successfulShards(), lessThan(restoreInfo.totalShards()));
+                assertEquals(numberOfPrimaryShards, restoreInfo.totalShards());
+            } catch (Exception e) {
+                assertThat(ExceptionsHelper.unwrapCause(e), instanceOf(ElasticsearchTimeoutException.class));
+            }
+        } finally {
+            for (MockTransportService transportService : transportServices) {
+                transportService.clearAllRules();
+            }
+
+            settingsRequest = new ClusterUpdateSettingsRequest();
+            TimeValue defaultValue = CcrSettings.INDICES_RECOVERY_ACTION_TIMEOUT_SETTING.getDefault(Settings.EMPTY);
+            settingsRequest.persistentSettings(Settings.builder().put(CcrSettings.INDICES_RECOVERY_ACTION_TIMEOUT_SETTING.getKey(),
+                defaultValue));
+            assertAcked(followerClient().admin().cluster().updateSettings(settingsRequest).actionGet());
         }
-
-
-        for (MockTransportService transportService : transportServices) {
-            transportService.clearAllRules();
-        }
-
-        settingsRequest = new ClusterUpdateSettingsRequest();
-        TimeValue defaultValue = CcrSettings.INDICES_RECOVERY_ACTION_TIMEOUT_SETTING.getDefault(Settings.EMPTY);
-        settingsRequest.persistentSettings(Settings.builder().put(CcrSettings.INDICES_RECOVERY_ACTION_TIMEOUT_SETTING.getKey(),
-            defaultValue));
-        assertAcked(followerClient().admin().cluster().updateSettings(settingsRequest).actionGet());
     }
 
     public void testFollowerMappingIsUpdated() throws IOException {

